Create profile loading mechanism

Bachmann1234 opened the pull request:
Allows users to register settings profiles and load them
in as needed. Allowing for different defaults for different
environments

Is this kind of what you were thinking for the profile setup?

This code should enable the following pattern

```
Settings.register_profile('dev', max_examples=5, max_shrinks=10, max_iterations=100)
Settings.register_profile('ci', max_examples=1000, max_shrinks=1000, max_iterations=2000)

Settings.load_profile(os.getenv('HYPOTHESIS_PROFILE', 'default'))
```

If you think im close ill iterate with you as needed. If we get to something you like let me know and ill write up the docs.

kxepal commented:
Nice feature, but I think you came from a wrong side. 

Assume you have multiple environments: dev, ci, stage, prod. For each you already have some configuration file. Otherwise there is no need in such division. And in these config files you may define hypothesis settings whatever you need and just let read them somewhere in `tests/__init__.py` to apply them before tests will get run. So you eventually don't need yet another configuration-per-environment thing.

Bachmann1234 commented:
@kxepal thats a good point. I guess I could just create the settings object on launch. 

The history here I wanted to set max_examples with an environment variable https://github.com/DRMacIver/hypothesis/pull/153 but then @DRMacIver pointed me to the idea of profiles he had http://www.drmaciver.com/2015/08/a-vague-roadmap-for-hypothesis-2-0/

But yeah, I guess since you can just define the settings when your test file gets loaded there may not be much of a need here. Maybe @DRMacIver had something more involved in mind for the `profiles` feature

DRMacIver commented:
So I still think this is a good feature, and I didn't really have anything much more involved in mind. :-)

It's not a very complicated feature, but having it makes peoples' lives easier and is useful for when you want to try something locally but faster.

In particular I think doing this in your test config basically requires multiple people to reinvent something that looks more or less like this. Additionally, I'd really like Hypothesis to ship with a bunch of built in profiles (I'm happy to write these @Bachmann1234, but you should feel to add any you like).

Anyway, thanks for doing this @Bachmann1234! This basically looks pretty good. I'll add some specific comments inline.

DRMacIver commented on src/hypothesis/settings.py:
```diff
@@ -257,8 +259,41 @@ def __exit__(self, *args, **kwargs):
 
-    default = DefaultSettings()
+    @staticmethod
+    def register_profile(name, **kwargs):
```
So in my head you would use explicit Settings objects here rather than kwargs, but I could be persuaded that kwargs are the way to go, or to support both modes. Is there a reason you wanted to do this as kwargs?

Bachmann1234 replied:
I was mostly trying to be more concise. But using the specific object would be more consistent.

Bachmann1234 replied:
Not to mention using the actual object is more in line with the whole Zen thing `Explicit is better than implicit.` :-P

The-Compiler commented:
What about having `HYPOTHESIS_PROFILE` interpreted by default?

DRMacIver commented on src/hypothesis/settings.py:
```diff
@@ -257,8 +259,41 @@ def __exit__(self, *args, **kwargs):
+        except KeyError:
+            raise ValueError("Profile '{}' has not been registered", name)
+        for key, value in profile.items():
+            setattr(Settings.default, key, value)
```
Assigning individual settings seems like an odd choice. In particular Settings.default is assignable and it seems worth taking advantage of that.

It would be nice for example to support things that had done:

```python
@given(blah, settings=Settings.default)
def test_stuff(blah):
    ...
```

not have Settings changed underneath them.

DRMacIver commented on src/hypothesis/settings.py:
```diff
@@ -257,8 +259,41 @@ def __exit__(self, *args, **kwargs):
+            setattr(Settings.default, name, setting.default)
 
+    @staticmethod
+    def load_profile(name):
```
Minor enhancement request: It would be nice to have a Settings.get_profile or similar method as well as one that loads it in by the default, so you could explicitly use a particular profile for one particular test.

DRMacIver commented:
I'm also strongly in favour of @The-Compiler's suggestion of having HYPOTHESIS_PROFILE interpreted by default.

Bachmann1234 commented:
Sound good. Read over the comments. Ill address them sometime early this week depending on my schedule.

Bachmann1234 commented:
So I addressed some of the comments but hit a snag.

I tried setting the default settings object rather then set things by hand.
```
@staticmethod
    def load_profile(name):
        """
        Loads in the settings defined in the profile provided
        If the profile does not exist an AttributeError will
        be thrown. Any setting not defined in the profile
        will be the library defined default for that setting
        """
        Settings.default = Settings.get_profile(name)
```

But this seems to break the context manager concept.

This test I added fails

```
def test_loading_profile_keeps_expected_behaviour():
    Settings.register_profile("ci", Settings(max_examples=10000))
    Settings.load_profile("ci")
    assert Settings().max_examples == 10000
    with Settings(max_examples=5):
        assert Settings().max_examples == 5
    assert Settings().max_examples == 10000
```

Any thoughts? What am I missing here?

I added the loading in a profile with an environment var btw. Testing it was a bit hacky though... See `test_load_on_environment` I could simply track the currently loaded profile... but I figured for now avoid the extra state.

Bachmann1234 commented:
Oh. I was thinking of adding a log when a profile is loaded in

"Profile 'ci' loaded"

DRMacIver commented:
How is the test with the context manager failing? It looks like it should work for me.

One thing that I think might be worth doing is making assigning Settings.default an error when you're inside a Settings context manager. It's... arguably a breaking change to the API, but it's so underdocumented that I'm not sure I could argue that the ability to do so is actually public, and I'm almost certain nobody is doing it and completely certain that it's not doing what they expect if they are.

DRMacIver commented:
Although I guess it's important that the ability to nest context managers works even if flat out assigning Settings.default doesn't.

DRMacIver commented:
If you do add a log I have a silly edge case question that I'm not sure what the best answer to is: What verbosity level do you use (I'd guess debug) and do you use the verbosity level of the new or the old settings value?

DRMacIver commented on src/hypothesis/settings.py:
```diff
@@ -257,8 +259,39 @@ def __exit__(self, *args, **kwargs):
+        try:
+            return Settings._profiles[name]
+        except KeyError:
+            raise ValueError("Profile '{}' has not been registered".format(name))
```
I would use InvalidArgument from hypothesis.errors here.

DRMacIver commented:
I guess on further thought there's no reason assignment inside a context manager can't just be a DeprecationWarning for now and keep the old and very broken behaviour in case anyone is actually using it. Has similar questions about which strict mode applies, but I think it makes sense for it to be the current one.

DRMacIver commented:
Extension to the feature that occurs to me that I'm going to record here but you should feel 100% free not to implement: It would be nice if the pytest plugin exposed a command line argument for setting the profile.

Bachmann1234 commented:
So here is how the test fails
```
/Users/bachmann/Code/hypothesis/tests/cover/test_settings.py:184: in test_loading_profile_keeps_expected_behaviour
    assert Settings().max_examples == 5
E   assert 10000 == 5
E    +  where 10000 = Settings(a_setting_just_for_these_tests=3, a_setting_with_limited_options=3, average_list_length=25.0, database_file='..., max_iterations=1000, max_shrinks=500, min_satisfying_examples=5, strict=True, timeout=-1, verbosity=Verbosity.normal).max_examples
E    +    where Settings(a_setting_just_for_these_tests=3, a_setting_with_limited_options=3, average_list_length=25.0, database_file='..., max_iterations=1000, max_shrinks=500, min_satisfying_examples=5, strict=True, timeout=-1, verbosity=Verbosity.normal) = Settings()
```

Basically, the option declared in the `with` statement is not being respected

Bachmann1234 commented:
After much digging I got the context manager working as expected

Bachmann1234 commented:
So as an update here is my TODO on this PR

1. See if my pytest option actually works as expected (looks like pytest stuff is tested by hand so ill just run some tests and put the output in here)
2. Squash https://github.com/Bachmann1234/hypothesis/commit/6ac683f7aa81c16e44fe2f9050f9eee0e8db9655 into https://github.com/Bachmann1234/hypothesis/commit/0d576dfa2fd6867398e2cad07c6899cdeb6283d3 Leaving them both here separately so you can compare the diff if you're inclined.
3. Write docs for this feature

As for the log I basically scrapped it after not being able to decide the best resolution to the verbosity question. To Log based on the previous or the new setting. 

Anything else im forgetting here?

The-Compiler commented:
> (looks like pytest stuff is tested by hand so ill just run some tests and put the output in here)

Judging from the amount of tests hypothesis has, I don't think anything is tested by hand - automate ALL the things! :wink:

See [tests/pytest](https://github.com/DRMacIver/hypothesis/tree/master/tests/pytest).

Bachmann1234 commented:
@The-Compiler AH! Thanks, ill update that either tomorrow or friday

DRMacIver commented on tests/cover/test_settings.py:
```diff
@@ -145,3 +147,53 @@ def test_can_assign_database(db):
+    # Monkeypatch the environment
+    monkeypatch.setenv("HYPOTHESIS_PROFILE", "ci")
+    with pytest.raises(hypothesis.errors.InvalidArgument):
+        # Reload. Non existent profile loaded so ValueError thrown
```
This comment is now wrong because it's an InvalidArgument thrown instead.

DRMacIver commented on src/hypothesis/settings.py:
```diff
@@ -261,11 +263,46 @@ def __exit__(self, *args, **kwargs):
+    @staticmethod
+    def register_profile(name, settings):
+        """
+        registers a collection of values to be used as a
```
Docstring now slightly strange because of the move from using kwargs to a Settings object.

DRMacIver commented:
Additional TODO items:

1. I've added a few trivial comments inline
2. The build is currently broken due to a mix of linting (you can fix this just by running tox -e lint and committing the changes) and possibly more serious issues (Settings has no attribute 'database'), which needs fixing.
3. You should definitely add yourself to the contributors list. :-)

DRMacIver commented:
BTW I've seen the 'Settings has no attribute 'database'' issue elsewhere and I'm pretty sure what's causing it is that you've got an AttributeError triggered inside the 'database' property. This causes Python to invoke the \_\_getattr\_\_ method (because Python attribute resolution is hilarious). The easiest way to debug what's going on is to call Settings.default.database directly and see what happens.

DRMacIver commented:
Note: If you rebase off master, #161 is now fixed and this is what was giving you trouble with the context manager. Assigning Settings.default should now work as expected.

Bachmann1234 commented:
Ok I think we are getting close here

Tests added to pytest plugin
Flake issues fixed
I fixed some of the build issues, ill wait to see what travis/appveyor says about that though
Added self to contributors list
Rebased from master and adjusted to account for the changes. 
Squashed down.

If that build smiles at me all I should just need to write up some docs and address any other feedback that comes up.

Bachmann1234 commented:
:-( the build frowns. More work to do

Bachmann1234 commented:
OK.. *NOW* I got all the lint issues fixed :-p this pushed me over the edge to get pyenv setup.

Long story short, hombrew installs python3.5, tox wants 3.4 for this, bla bla bla. Now I got my python environment cleaned up some. Next to fix some tests failing for reasons that are unclear to me... but its getting late.

Bachmann1234 commented:
But yeah, still gotta account for settings defined outside settings.py

I thought updating the profiles and the current default with the new property would be enough but testing that locally showed the issue was not fixed so I did not commit it. Ill have to come back to this again. Sorry for the PR chrun. I honestly thought this would be a bit easier :-P

DRMacIver commented:
Sorry. Welcome to my life. :-) Hypothesis isn't the easiest project to work on and the settings system is a legacy of when I was being, uh, "clever" early in the project's life.

RE settings defined outside of settings.py: If it will make your life easier you should feel free to move all setting definitions into settings.py and then freeze settings in some way so that define_setting can't be called afterwards. The need for it is basically gone (it was necessitated by the old and terrible strategy API), and it was never part of the public API.

Bachmann1234 commented:
Fixed issues with the coverage tests and took a first draft for the docs. While writing this I ran into a bit of a issue with the environment variable idea. 

To define profiles I need to import settings, which already loads in the profile from the environment. 

So to address this I added a function that loads the profile defined in the environment. Though honestly it may be better to remove this function and just document this pattern for loading in a profile from the environment.

```
>>> from hypothesis import Settings
>>> Settings.register_profile("ci", Settings(max_examples=1000))
>>> Settings.register_profile("dev", Settings(max_examples=10))
>>> Settings.register_profile("debug", Settings(max_examples=10, verbosity=Verbosity.verbose))
>>> Settings.load(os.getenv(u'HYPOTHESIS_PROFILE', 'default'))
```

Bachmann1234 commented:
Here are docs draft rendered
![screen shot 2015-09-28 at 12 21 09 am](https://cloud.githubusercontent.com/assets/1521093/10127874/439a793a-6577-11e5-89c1-d28d937bd16b.png)

Bachmann1234 commented:
Here is the draft rendered.
![screen shot 2015-09-28 at 12 28 02 am](https://cloud.githubusercontent.com/assets/1521093/10127897/d166f70c-6577-11e5-93fa-1894641afa98.png)

Bachmann1234 commented:
Here is a draft of the rendered docs
![screen shot 2015-09-28 at 12 28 02 am](https://cloud.githubusercontent.com/assets/1521093/10127906/efb88e78-6577-11e5-9282-e1f71b1fb930.png)

Bachmann1234 commented on src/hypothesis/settings.py:
```diff
@@ -259,6 +266,52 @@ def __exit__(self, *args, **kwargs):
+            return copy.copy(Settings._profiles[name])
+        except KeyError:
+            raise InvalidArgument(
+                "Profile '{}' has not been registered".format(
```
This is python 2.7 style formatting. I need to update it to work for 2.6

DRMacIver replied:
Yeah, sorry. :-( I very much regret supporting Python 2.6 but each little change like this doesn't seem *quite* worth dropping it for.

DRMacIver commented:
So here's a possible solution to the problem with the environment variable: Make Settings.default lazily initialized and check for the profile from the environment variable being defined the first time you access it (so that you're implicitly calling get_profile(os.environ.get('HYPOTHESIS_PROFILE')) the first time you access Settings.default.

Note: This may be super complicated and you really should feel free to ditch the environment variable feature if you struggle to make it work.

DRMacIver commented:
The docs look good BTW.

Bachmann1234 commented:
Ended up scrapping the environment variable but left the pattern in the docs. I just was not seeing a clear   path to fixing the issue.

<img width="1086" alt="screen shot 2015-09-28 at 9 55 47 pm" src="https://cloud.githubusercontent.com/assets/1521093/10154681/7f77353a-6638-11e5-84a4-300dcc54e836.png">

DRMacIver commented on src/hypothesis/settings.py:
```diff
@@ -198,6 +201,10 @@ def define_setting(cls, name, description, default, options=None):
         setattr(cls, name, SettingsProperty(name))
+        if cls.default:
+            setattr(cls.default, name, default)
+        for profile in cls._profiles.values():
```
This doesn't look correct to me. What happens if I do:

```python
x = Settings()
define_setting('blah', default=3, ...)
Settings.register_profile('hi', Settings(blah=2))
assert x.blah == 3
assert Settings.get_profile('hi') == 3
```

Bachmann1234 replied:
Ok so I added a test 
```
def test_define_setting_then_loading_profile():
    x = Settings()
    Settings.define_setting(
        u'fun_times',
        default=3, description=u'Something something spoon',
        options=(1, 2, 3, 4),
    )
    Settings.register_profile('hi', Settings(fun_times=2))
    assert x.fun_times == 3
    assert Settings.get_profile('hi').fun_times == 2
```

So I make a settings object `x`

I define a new setting 

x gets the default value while any new profiles get the value I set. Seems right to me. Though this PR is the lat thing I do before bed so I would not be surprised if i'm just being dense

DRMacIver replied:
I'm very confused what's going on here, but this PR is the first thing I do when drinking my coffee so I also would not be surprised if I were being dense. :-)

Why is the setattr(profile, name, default) not clobbering this? There must be an attribute lookup happening before the assignment I guess, but relying on that seems very suspect.

OTOH with the test there I don't mind too much. The settings system is due for a significant rework and I'm happy to get this merged and do the internals cleanup myself.

Bachmann1234 replied:
honestly im not 100% sure. You mentioned attribute lookup. When I register/load profiles I perform a copy so that may do it

DRMacIver commented on docs/details.rst:
```diff
@@ -285,6 +285,55 @@ You can also override the default by setting the environment variable
+~~~~~~~~~~~~~~~~~
+Settings Profiles
+~~~~~~~~~~~~~~~~~
+
```
The documentation should also mention that there's an argument you can use if you're using the pytest plugin.

DRMacIver commented:
The current set of failing tests is almost entirely novel, but I don't see how it can be your fault. Don't worry about them for now, I'll look into it. Yet another battle in the endless war of Hypothesis build stability. Sorry.

Bachmann1234 commented:
<img width="751" alt="screen shot 2015-09-29 at 11 35 17 pm" src="https://cloud.githubusercontent.com/assets/1521093/10183879/cc718be0-6702-11e5-87a7-4778b8736c03.png">
<img width="755" alt="screen shot 2015-09-29 at 11 35 01 pm" src="https://cloud.githubusercontent.com/assets/1521093/10183880/cc7455be-6702-11e5-8c18-c92a69a4c367.png">

DRMacIver commented:
Sorry, the ascii failures are a problem that's worked around in master (It's a bug in py.test 2.8.1).

DRMacIver commented:
That is, if you rebase this off master they should go away.

DRMacIver commented:
I think once you've rebased off master to fix the pytest issues this LGTM. Are you aware of anything else you need to work on before merging?

Bachmann1234 commented:
@DRMacIver rebased. Lets see what the build says

Bachmann1234 commented:
Timeout? That's new

DRMacIver commented:
A timeout is usually just Travis being under load. I've rebooted the job that failed and will rerun it, but I'm pretty sure it will pass and then I will merge. \o/

Bachmann1234 commented:
Woot!

DRMacIver commented:
:shipit:  Thanks so much for all your work on this. Sorry it took so long.

Bachmann1234 commented:
No problem!

Bachmann1234 commented:
@DRMacIver if any issues come up around this dont be afraid to reach out

The-Compiler commented on docs/details.rst:
```diff
@@ -285,6 +285,64 @@ You can also override the default by setting the environment variable
+    >>> Settings.register_profile("debug", Settings(max_examples=10, verbosity=Verbosity.verbose))
+    >>> Settings.load_profile(os.getenv(u'HYPOTHESIS_PROFILE', 'default'))
+
+If you are using the hypothesis pytest plugin. If the profile was registered
```
These two sentences are confusing me - shouldn't this be something like "If you are using the hypothesis pytest plugin and the profile ..."?

Bachmann1234 replied:
Sure, I can edit that tonight. and open a quick PR

The-Compiler commented on src/hypothesis/extra/pytestplugin.py:
```diff
@@ -19,6 +19,9 @@
 PYTEST_VERSION = tuple(map(int, pytest.__version__.split('.')[:3]))
+LOAD_PROFILE_OPTION = '--hypothesis-profile'
+
+PYTEST_VERSION = tuple(map(int, pytest.__version__.split('.')))
```
That addition seems wrong, as it's already defined above (with the `[:3]` introduced recently)

DRMacIver replied:
Yeah, I noticed this right after merging. It's now fixed in master. I think this was a hiccup caused by the last rebase.

The-Compiler commented:
Sorry I didn't look at this before the merge - but the two things above (especially the second one) probably need some amending :wink:
